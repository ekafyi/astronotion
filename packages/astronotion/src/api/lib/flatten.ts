import type { BlockMap, Decoration } from "notion-types";

/**
 * Get plain text from a notion-client "Block" data, which consists
 * of nested arrays with formatting metadata.
 *
 * @param {Decoration[]} data
 * @returns {string} text content
 */
export const flattenText = (data: Decoration[]) =>
	data
		.map((item) => item[0])
		.flat()
		.join(" ")
		.replace(/\s\s+/g, " ");

export const flattenBlocks = (blockMap: BlockMap) =>
	Object.entries(blockMap).map((item) => item[1].value);
