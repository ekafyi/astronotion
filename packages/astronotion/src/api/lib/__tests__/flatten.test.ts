import sourceTwo from "../../../../fixtures/notion-client-entry-page-1.json";
import sourceOne from "../../../../fixtures/notion-client-parent-page-1.json";

import { flattenText, flattenBlocks } from "../flatten";

const decoratedTextOne = [["Journal"]];
const decoratedTextTwo = [
	["Click "],
	["All Entries", [["c"]]],
	[" to filter entries by a specific category such as "],
	["daily", [["b"]]],
	[" or "],
	["personal", [["b"]]],
	["."],
];

const blocksOne = [
	sourceOne.block["2ba80ec3-d84d-4647-9f23-ec15ba5e39b0"],
	sourceOne.block["0588e349-5e26-40f0-b791-54618c4f285b"],
];
const blocksTwo = [
	sourceTwo.block["0588e349-5e26-40f0-b791-54618c4f285b"],
	sourceTwo.block["14c6292d-f91f-4302-9854-7932dbd419a9"],
];

describe("flatten text", () => {
	test("flatten simple text", () => {
		// @ts-expect-error using static JSON data source
		expect(flattenText(decoratedTextOne)).toMatchInlineSnapshot(`"Journal"`);
	});
	test("flatten formatted text", () => {
		// @ts-expect-error using static JSON data source
		expect(flattenText(decoratedTextTwo)).toMatchInlineSnapshot(
			`"Click All Entries to filter entries by a specific category such as daily or personal ."`
		);
	});
});

describe("convert blocks to array", () => {
	test("from parent page", () => {
		// @ts-expect-error using static JSON data source
		expect(flattenBlocks(blocksOne)).toMatchSnapshot();
	});
	test("from entry/child page", () => {
		// @ts-expect-error using static JSON data source
		expect(flattenBlocks(blocksTwo)).toMatchSnapshot();
	});
});
