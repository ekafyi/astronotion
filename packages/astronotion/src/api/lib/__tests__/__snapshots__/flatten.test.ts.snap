// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert blocks to array from entry/child page 1`] = `
Array [
  Object {
    "alive": true,
    "created_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "created_by_table": "notion_user",
    "created_time": 1651059000000,
    "id": "0588e349-5e26-40f0-b791-54618c4f285b",
    "last_edited_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "last_edited_by_table": "notion_user",
    "last_edited_time": 1651059000000,
    "parent_id": "1cf6f82e-ffc2-4ad4-aea5-a50f88198a7b",
    "parent_table": "block",
    "space_id": "f2ae8a7d-484c-4fc5-aca1-187de317cdf1",
    "type": "text",
    "version": 5,
  },
  Object {
    "alive": true,
    "content": Array [
      "0f37dc37-b7f0-4adb-a7ad-2f991851cb8b",
      "9a023b03-6e7d-43f9-9d52-ea5387c1bd12",
    ],
    "created_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "created_by_table": "notion_user",
    "created_time": 1651923180000,
    "id": "14c6292d-f91f-4302-9854-7932dbd419a9",
    "last_edited_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "last_edited_by_table": "notion_user",
    "last_edited_time": 1651923180000,
    "parent_id": "1cf6f82e-ffc2-4ad4-aea5-a50f88198a7b",
    "parent_table": "block",
    "space_id": "f2ae8a7d-484c-4fc5-aca1-187de317cdf1",
    "type": "column_list",
    "version": 5,
  },
]
`;

exports[`convert blocks to array from parent page 1`] = `
Array [
  Object {
    "alive": true,
    "collection_id": "fa3d5bda-3e25-46e4-b86c-3157329bad83",
    "copied_from": "aebe0ee9-b0df-4ebb-bc61-ed69a09c7c8e",
    "created_by": "2dcbe436-2aba-45bb-a3d4-a245488b637d",
    "created_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "created_by_table": "notion_user",
    "created_time": 1650189589678,
    "format": Object {
      "block_locked": false,
      "block_locked_by": "d0a8cba1-b998-46a3-bcd1-0efb420410b1",
      "collection_pointer": Object {
        "id": "fa3d5bda-3e25-46e4-b86c-3157329bad83",
        "spaceId": "f2ae8a7d-484c-4fc5-aca1-187de317cdf1",
        "table": "collection",
      },
      "copied_from_pointer": Object {
        "id": "aebe0ee9-b0df-4ebb-bc61-ed69a09c7c8e",
        "spaceId": "1f88cc90-92fd-4ce4-bfcd-25daec2ffbbe",
        "table": "block",
      },
      "page_cover_position": 0.5,
    },
    "id": "2ba80ec3-d84d-4647-9f23-ec15ba5e39b0",
    "last_edited_by": "c62621c6-e8ec-40f6-9ab7-334f43db12b4",
    "last_edited_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "last_edited_by_table": "notion_user",
    "last_edited_time": 1651934220000,
    "parent_id": "f2ae8a7d-484c-4fc5-aca1-187de317cdf1",
    "parent_table": "space",
    "permissions": Array [
      Object {
        "role": "editor",
        "type": "user_permission",
        "user_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
      },
      Object {
        "bot_id": "629a6b0c-dbfe-47d4-a5f5-41cc722d4e20",
        "parent_id": "f2ae8a7d-484c-4fc5-aca1-187de317cdf1",
        "parent_table": "space",
        "role": Object {
          "read_content": true,
        },
        "type": "bot_permission",
      },
      Object {
        "added_timestamp": 1651934263790,
        "allow_duplicate": false,
        "role": "reader",
        "type": "public_permission",
      },
    ],
    "space_id": "f2ae8a7d-484c-4fc5-aca1-187de317cdf1",
    "type": "collection_view_page",
    "version": 50,
    "view_ids": Array [
      "78b3a536-90ea-4d1b-939e-cc443914994e",
      "5576426d-d9a5-4501-a4b0-db414aadf15e",
      "410d77ec-e304-4567-858d-3ba2218b4656",
      "78586561-4506-4d86-b2d9-2de435265e7b",
    ],
  },
  Object {
    "alive": true,
    "created_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "created_by_table": "notion_user",
    "created_time": 1651059000000,
    "id": "0588e349-5e26-40f0-b791-54618c4f285b",
    "last_edited_by_id": "cdfe101e-2758-4d24-ac9c-b62ab4f7e9c2",
    "last_edited_by_table": "notion_user",
    "last_edited_time": 1651059000000,
    "parent_id": "1cf6f82e-ffc2-4ad4-aea5-a50f88198a7b",
    "parent_table": "block",
    "space_id": "f2ae8a7d-484c-4fc5-aca1-187de317cdf1",
    "type": "text",
    "version": 5,
  },
]
`;
